crates/client/src/lib.rs

```diff
@@ -574,71 +557,169 @@ impl<R: RegistryStorage, C: ContentStorage, N: NamespaceMapStorage> Client<R, C,
+            let response = match self
                 .api
-                .fetch_logs(FetchLogsRequest {
-                    log_length: checkpoint.log_length,
-                    operator: operator
-                        .head_fetch_token
-                        .as_ref()
-                        .map(|t| Cow::Borrowed(t.as_str())),
-                    limit: None,
-                    packages: Cow::Borrowed(&last_known),
-                })
+                .fetch_logs(
+                    registry_domain,
+                    FetchLogsRequest {
+                        log_length: checkpoint.log_length,
+                        operator: operator
+                            .head_fetch_token
+                            .as_ref()
+                            .map(|t| Cow::Borrowed(t.as_str())),
+                        limit: None,
+                        // last known fetch token for each package log ID
+                        packages: Cow::Owned(
+                            packages
+                                .iter()
+                                .map(|(id, p)| (id.clone(), p.head_fetch_token.clone()))
+                                .collect::<IndexMap<_, _>>(),
+                        ),
+                    },
+                )
                 .await
                 .inspect(|res| {
                     for warning in res.warnings.iter() {
                         tracing::warn!("Fetch warning from registry: {}", warning.message);
                     }
-                })
-                .map_err(|e| {
-                    ClientError::translate_log_not_found(e, |id| {
-                        packages.get(id).map(|p| p.name.clone())
-                    })
-                })?;
+                }) {
+                Ok(res) => Ok(res),
+                Err(err) => match &err {
+                    api::ClientError::Fetch(FetchError::LogNotFound(log_id))
+                    | api::ClientError::Package(PackageError::LogNotFound(log_id)) => {
+                        if let Some(name) = packages.get(log_id).map(|p| p.name.clone()) {
+                            Err(ClientError::PackageDoesNotExist { name })
+                        } else {
+                            Err(ClientError::Api(err))
+                        }
+                    }
+                    api::ClientError::LogNotFoundWithHint(log_id, hint) => {
+                        match hint.to_str().ok().map(|s| s.split_once('=')) {
+                            Some(Some((namespace, registry)))
+                                if !self.ignore_federation_hints
+                                    && packages.contains_key(log_id) =>
+                            {
+                                let package_name = &packages.get(log_id).unwrap().name;
+
+                                if self.auto_accept_federation_hints
+                                    || Confirm::with_theme(&ColorfulTheme::default())
```

Comment:
Yeah, this crate is only being used for CLIs at the moment. I was leaning towards simplifying this crate for that use case.

Reply:
@lann Fine as is? Feature flag?

If refactored into a callback, it would need be `async`. Started looking complex and wasn't sure it was worthwhile.

Reply:
It can be refactored later.